File: lib_base/src/main/java/com/luckyaf/smartandroid/lib/base/app/BaseFragment.java
package com.luckyaf.smartandroid.lib.base.app;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.luckyaf.smartandroid.lib.base.utils.ToastUtil;

/**
 * 类描述：
 *
 * @author Created by luckyAF on 2018/7/6
 */
public abstract  class BaseFragment extends Fragment {

    private static String TAG = BaseFragment.class.getName();
    protected Context mContext;
    protected Activity mActivity;

    /**
     * 根view
     */
    protected View rootView;


    /**
     * 是否使用懒加载
     */
    protected boolean mLazyLoad = false;


    /**
     * 首次显示
     */
    private boolean isFirstResumeVisible = true;
    /**
     * 首次显示
     */
    private boolean isPrepared = false;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        this.mContext = context;
        this.mActivity = (Activity) context;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container
            , Bundle savedInstanceState) {
        if (null == rootView) {
            rootView = inflater.inflate(getLayoutId(), container, false);
        }
        initParams(getArguments());
        rootView.setClickable(true);
        return rootView;
    }

    @SuppressWarnings("unchecked")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initView(view, savedInstanceState);
        isPrepared = true;
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (!mLazyLoad) {
            initData();
        }
    }





    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    /**
     * 该抽象方法就是 onCreateView中需要的layoutID
     *
     * @return fragment layout id
     */
    protected abstract @LayoutRes
    int getLayoutId();

    /**
     * 初始化参数
     *
     * @param bundle 参数
     */
    protected abstract void initParams(Bundle bundle);

    /**
     * 该抽象方法就是 初始化view
     *
     * @param view               初始化view
     * @param savedInstanceState activity传来的数据
     */
    protected abstract void initView(View view, Bundle savedInstanceState);

    /**
     * 数据初始化
     */
    protected abstract void initData();



    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
        if (isVisibleToUser) {
            if (mLazyLoad && isFirstResumeVisible && isPrepared) {
                initData();
                isFirstResumeVisible = false;
            }
            onVisibleToUser();
        } else {
            onInVisibleToUser();
        }
    }

    /**
     * 对用户可见
     */
    protected void onVisibleToUser() {

    }


    /**
     * 对用户不可见
     */
    protected void onInVisibleToUser() {

    }


    @SuppressWarnings("unchecked")
    protected <V extends View> V findView(@IdRes int resId) {
        return (V) rootView.findViewById(resId);
    }

    public void showMessage(String message) {
        ToastUtil.showMessage(mContext,message);
    }

    public void showMessage(@StringRes int message) {
        ToastUtil.showMessage(mContext,message);
    }


}

File: module_kotlin/src/main/java/com/luckyaf/oneforall/kotlin/base/BaseFragment.kt
package com.luckyaf.oneforall.kotlin.base

import android.content.Context
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.luckyaf.oneforall.kotlin.utils.ToastUtil
import io.reactivex.disposables.Disposable

/**
 * 类描述：fragment基类
 * @author Created by luckyAF on 2018/7/18
 *
 */
@Suppress("unused")
abstract class BaseFragment<T : BasePresenter> : Fragment(), BaseView {

    lateinit var mPresenter : T


    lateinit var mContext: Context

    /**
     * 是否使用懒加载
     */
    protected var mLazyLoad = false

    /**
     * 首次显示
     */
    private var isFirstResumeVisible = true
    /**
     * 是否准备好
     */
    private var isPrepared = false

    lateinit var mRootView: View

    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.mContext = context
    }

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        mRootView = inflater.inflate(getLayoutId(), container, false)
        initParams(arguments)
        mPresenter = getPresenter()
        return mRootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initView()
        isPrepared = true
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (!mLazyLoad) {
            initData()
        }
    }


    /**
     * 加载布局
     */
    @LayoutRes
    abstract fun getLayoutId(): Int

    /**
     * 获取presenter
     */
    abstract fun getPresenter(): T


    /**
     * 初始化参数
     */
    abstract fun initParams(bundle: Bundle?)

    /**
     * 初始化 ViewI
     */
    abstract fun initView()

    /**
     * 初始化数据
     */
    abstract fun initData()


    fun <T : View>findView(viewId: Int): T{
        return mRootView.findViewById(viewId)
    }


    fun addSubscription(disposable: Disposable?) {
        if (null != disposable) {
            if (mPresenter is RxPresenter) {
                (mPresenter as RxPresenter).addSubscription(disposable)
            }
        }
    }

    fun removeSubscription(disposable: Disposable?) {
        if (null != disposable) {
            if (mPresenter is RxPresenter) {
                (mPresenter as RxPresenter).removeSubscription(disposable)
            }
        }
    }


    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        if (isVisibleToUser) {
            if (mLazyLoad && isFirstResumeVisible && isPrepared) {
                initData()
                isFirstResumeVisible = false
            }
            onVisibleToUser()
        } else {
            onInVisibleToUser()
        }
    }

    /**
     * 对用户可见
     */
    protected fun onVisibleToUser() {

    }


    /**
     * 对用户不可见
     */
    protected fun onInVisibleToUser() {

    }


    override fun onDestroy() {
        mPresenter.detachView()
        super.onDestroy()
    }


    override fun showMessage(message: Int) {
        //ToastUtil.showMessage(mContext, message)
        view?.let {
            Snackbar.make(it, message, Snackbar.LENGTH_SHORT).show()
        }

    }

    override fun showMessage(message: String) {
        //ToastUtil.showMessage(mContext, message)
        view?.let {
            Snackbar.make(it, message, Snackbar.LENGTH_SHORT).show()
        }
    }

    override fun showProgress() {
        //
    }

    override fun showProgress(message: String) {
    }

    override fun showProgress(message: Int) {
    }

    override fun hideProgress() {
        //
    }

}